Signal ALPN/SNI availability during handshake

Ralith opened the pull request:
Fixes #749. Will improve our performance on the quic-tracker interop tests.

djc commented on quinn/src/connection.rs:
```diff
@@ -38,12 +38,41 @@ where
 {
-    pub(crate) fn new(conn: ConnectionRef<S>, connected: oneshot::Receiver<bool>) -> Self {
-        Self {
+    pub(crate) fn new(
```
I'm fully on board with encapsulating the coupling of the driver and `Connecting`. However, this change appears to closely couple `Connecting` and the `ConnectionRef`, and this seems unnecessary to achieve the former goal (we could still make `ConnectionRef::new()` private to the `connection` module, of course).

Ralith replied:
I'm a bit concerned about the growth of the `ConnectionRef::new` arguments list, but it's not untenable just yet. Applied.

djc commented on quinn-proto/src/crypto/rustls.rs:
```diff
@@ -217,16 +219,16 @@ pub struct AuthenticationData {
-    ///
-    /// If this is `None`, and `Connection::is_handshaking` returns `false`, the connection
-    /// will have already been closed.
+    /// Available while fully connected, if provided by the peer.
```
So I feel like it's a bit ugly that the future resolving does not have much bearing on the availability of the `peer_certificates`. Should we separate them out after all?

Ralith replied:
I agree that this is a bit ugly. I'm uncertain what a more fine-grained version should look like. The most obvious approach is just split `AuthenticationData` into `EarlyAuthenticationData` and `LateAuthenticationData`, but that's awfully abstract/obscure. We could return to hardcoded interfaces for SNI and ALPN and leave `AuthenticationData` for the certificates; I think this is a bit more appealing, as stubbing out SNI/ALPN when unavailable doesn't seem any worse than having a weird empty `EarlyAuthenticationData`.

djc replied:
I agree that an early/late split isn't particularly appealing. However, thinking about it more, the current approach with the ALPN protocol is also wrong. As I understand, applications are not required to use ALPN, in which case this future doesn't actually ever wake (I think).

"TLS uses Application Layer Protocol Negotiation (ALPN) to select an application protocol. Unless another mechanism is used for agreeing on an application protocol, endpoints MUST use ALPN for this purpose." (-tls, section 8.1)

As such, ideally we should get insight from rustls on how far along the handshake is. I suppose the alternative solution is that we wake the task up when the handshake terminates.

For just certificates, the name `AuthenticationData` feel overly broad. In that case, we might just call it `PeerIdentity` or some such.

Ralith replied:
> applications are not required to use ALPN

Whoops, good catch. I don't think SNI is required either, so we indeed need a new signal from rustls here. Something like the recent proposal in https://github.com/ctz/rustls/issues/89 would trivialize this, but a lower-impact solution would be a new state accessor in the QUIC-specific API.

Ralith replied:
> I suppose the alternative solution is that we wake the task up when the handshake terminates.

That would defeat the goal of allowing ALPN-dependent 0-RTT data to be processed immediately.

> For just certificates, the name AuthenticationData feel overly broad. In that case, we might just call it PeerIdentity or some such.

Yeah, that makes sense.

djc replied:
> That would defeat the goal of allowing ALPN-dependent 0-RTT data to be processed immediately.

Well, you can still optimize for the case where ALPN is provided by the client, right? Which it's easy for any protocol that wants the performance to opt in to, supposedly.

Ralith replied:
Oh, I see what you mean. Yeah, that'd be a decent short-term solution in lieu of a rustls change.

Ralith commented:
I've rebased and split `AuthenticationData` into separate `HandshakeData` and `Identity` structures, and focused the async notification (which is now reliably delivered) around the former. Note that github is misordering the commits fairly severely.
